include/exadg/incompressible_navier_stokes/spatial_discretization/viscosity_model.cpp

```diff
@@ -0,0 +1,838 @@
+void
+ViscosityModel<dim, Number, use_turbulence_model, use_generalized_newtonian_model>::
+  calculate_viscosity(VectorType const & velocity) const
+{
+  VectorType dummy;
+
+  matrix_free->loop(&This::cell_loop_set_coefficients,
+                    &This::face_loop_set_coefficients,
+                    &This::boundary_face_loop_set_coefficients,
+                    this,
+                    dummy,
+                    velocity);
+}
+
+template<int dim, typename Number, bool use_turbulence_model, bool use_generalized_newtonian_model>
+void
+ViscosityModel<dim, Number, use_turbulence_model, use_generalized_newtonian_model>::
+  cell_loop_set_coefficients(dealii::MatrixFree<dim, Number> const & matrix_free,
+                             VectorType &,
+                             VectorType const & src,
+                             Range const &      cell_range) const
+{
+  CellIntegratorU integrator(matrix_free,
+                             turbulence_model_data.dof_index,
+                             turbulence_model_data.quad_index);
+
+  // containers needed dependent on template parameters
+  scalar filter_width;
+  scalar viscosity;
+
+  // loop over all cells
+  for(unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
+  {
+    integrator.reinit(cell);
+    integrator.read_dof_values(src);
+
+    // we only need the gradient
+    integrator.evaluate(dealii::EvaluationFlags::gradients);
+
+    // get filter width for this cell
+    if(use_turbulence_model == true)
+    {
+      filter_width = integrator.read_cell_data(this->filter_width_vector);
+    }
+
+    // loop over all quadrature points
+    for(unsigned int q = 0; q < integrator.n_q_points; ++q)
+    {
+      // calculate needed quantities for this cell
+      tensor velocity_gradient           = integrator.get_gradient(q);
+      tensor symmetric_velocity_gradient = dealii::make_vectorized_array<Number>(0.5) *
+                                           (velocity_gradient + transpose(velocity_gradient));
+      scalar shear_rate_squared =
+        dealii::make_vectorized_array<Number>(2.0) *
+        scalar_product(symmetric_velocity_gradient, symmetric_velocity_gradient);
+
+      if(use_generalized_newtonian_model == false)
+      {
+        viscosity =
+          dealii::make_vectorized_array<Number>(turbulence_model_data.kinematic_viscosity);
+      }
+      else
+      {
+        set_generalized_newtonian_viscosity(shear_rate_squared, viscosity);
+      }
+
+      if(use_turbulence_model == true)
+      {
+        add_turbulent_viscosity(viscosity /*might use generalized Newtonian viscosity*/,
+                                filter_width,
+                                velocity_gradient,
+                                symmetric_velocity_gradient,
+                                shear_rate_squared,
+                                turbulence_model_data.constant);
+      }
```

Comment:
I currently believe we should think about a more modular design where different models (living in `IncNS::SpatialOperatorBase`, e.g. `TurbulenceModel`, `GeneralizedNewtonian...`) may accumulate/assemble into the same `VariableCoefficients` field. This implies slightly more computational work. However, I otherwise have the concern that things get coupled too strongly in the code, making changes and maintenance more and more challenging.